project Add ShoppingCart model is ready for the code review

misterSokor opened the pull request.

misterSokor committed:
project Add ShoppingCart model is ready for the code review

Elena-Bruyako commented on src/main/java/com/store/litflix/controller/CartController.java:
```diff
@@ -0,0 +1,72 @@
+    )
+    @PostMapping
+    public ShoppingCartResponseDto addBookToCart(
+            @RequestBody ShoppingCartRequestDto request) {
```
```suggestion
            @RequestBody @Valid ShoppingCartRequestDto request) {
```

Elena-Bruyako replied:
not fixed

Elena-Bruyako commented on src/main/java/com/store/litflix/controller/CartController.java:
```diff
@@ -0,0 +1,72 @@
+    @PutMapping("/cart-items/{cartItemId}")
+    public ShoppingCartResponseDto updateCartItemQuantity(
+            @PathVariable Long cartItemId,
+            @RequestBody UpdateCartItemQuantityRequestDto requestDto
```
```suggestion
            @RequestBody @Valid UpdateCartItemQuantityRequestDto requestDto
```

Elena-Bruyako replied:
not fixed

Elena-Bruyako commented on src/main/java/com/store/litflix/dto/cart/ShoppingCartRequestDto.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class ShoppingCartRequestDto {
```
```suggestion
public class CartItemRequestDto {
```

Elena-Bruyako replied:
rename and validate fields

Elena-Bruyako commented on src/main/java/com/store/litflix/dto/cart/UpdateCartItemQuantityRequestDto.java:
```diff
@@ -0,0 +1,8 @@
+import lombok.Data;
+
+@Data
+public class UpdateCartItemQuantityRequestDto {
```
```suggestion
public class UpdateCartItemDto {
```

Elena-Bruyako commented on src/main/java/com/store/litflix/dto/cart/UpdateCartItemQuantityRequestDto.java:
```diff
@@ -0,0 +1,8 @@
+
+@Data
+public class UpdateCartItemQuantityRequestDto {
+    private int quantity;
```
add validation

Elena-Bruyako commented on src/main/java/com/store/litflix/exception/CartItemAccessDeniedException.java:
```diff
@@ -0,0 +1,7 @@
+package com.store.litflix.exception;
+
+public class CartItemAccessDeniedException extends RuntimeException {
```
remove

Elena-Bruyako commented on src/main/java/com/store/litflix/exception/CartItemNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package com.store.litflix.exception;
+
+public class CartItemNotFoundException extends RuntimeException {
```
remove

Elena-Bruyako commented on src/main/java/com/store/litflix/model/CartItem.java:
```diff
@@ -0,0 +1,35 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotNull
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id")
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

Elena-Bruyako replied:
Why did you react but not change anything?

Elena-Bruyako commented on src/main/java/com/store/litflix/model/CartItem.java:
```diff
@@ -0,0 +1,35 @@
+    private ShoppingCart shoppingCart;
+    @NotNull
+    @ManyToOne
+    @JoinColumn(name = "book_id")
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

Elena-Bruyako replied:
same here

Elena-Bruyako commented on src/main/java/com/store/litflix/model/ShoppingCart.java:
```diff
@@ -0,0 +1,39 @@
+@Entity
+@Getter
+@Setter
+@ToString
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/litflix/model/ShoppingCart.java:
```diff
@@ -0,0 +1,39 @@
+@SQLDelete(sql = "UPDATE carts SET is_deleted = true WHERE id = ?")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/litflix/model/ShoppingCart.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ManyToOne(optional = false)
+    @JoinColumn(name = "user_id", nullable = false)
+    @ToString.Exclude
```
```suggestion
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false)
```

Elena-Bruyako commented on src/main/java/com/store/litflix/repository/cart/CartItemRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long>,
+        JpaSpecificationExecutor<CartItem> {
```
```suggestion
public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```

Elena-Bruyako replied:
remove `JpaSpecificationExecutor<CartItem>`

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+    public ShoppingCartResponseDto addBookToCart(ShoppingCartRequestDto requestDto) {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = userRepository.findByEmail(authentication.getName())
+                .orElseThrow(() -> new RuntimeException("User not found"));
```
provide parameter to message
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("User not found"));
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+    @Override
+    public ShoppingCartResponseDto addBookToCart(ShoppingCartRequestDto requestDto) {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = userRepository.findByEmail(authentication.getName())
```
don't use name or email, find user by userId

misterSokor replied:
Hi Elena, could you please clarify your suggestion? In my current version of JwtAuthenticationFilter, the user is stored in the SecurityContext using their email, password, and roles (via UserDetails). Since the email is used to identify the user, I'm not sure how I could use a userId in the service layer without modifying the filter or token structure. How would you recommend accessing the userId in this case, or was the idea to update the JWT to store and extract the ID instead?

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+                .orElseThrow(() -> new RuntimeException("User not found"));
+
+        Book book = bookRepository.findById(requestDto.getBookId()).orElseThrow(
+                () -> new RuntimeException("The book wos not found.")
```
EntityNotFoundException, parameter

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+                () -> new RuntimeException("The book wos not found.")
+        );
+
+        int quantity = requestDto.getQuantity();
```
why do you need this variable?

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+
+        ShoppingCart shoppingCart = cartRepository.findByUserId(user.getId())
+                .orElseGet(() -> {
+                    ShoppingCart newCart = new ShoppingCart();
```
create separate method `createShoppingCart()` and call it during the registration

misterSokor replied:
I thought that if we create a shopping cart at registration, it might just occupy unnecessary space in the database, especially if the user doesn't plan on purchasing anything. Maybe it's better to create the cart only when the user wants to add something to it?

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+
+        ShoppingCart saved = cartRepository.save(shoppingCart);
+
+        return shoppingCartMapper.toDto(saved);
```
```suggestion
        cartRepository.save(shoppingCart);
        return shoppingCartMapper.toDto(shoppingCart);
```

Elena-Bruyako commented on src/main/resources/application.properties:
```diff
@@ -11,5 +11,8 @@ spring.jpa.show-sql=true
 
+server.error.include-message=always
+server.error.include-stacktrace=never
+
```
```suggestion 
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+        User user = userRepository.findByEmail(auth.getName())
+                .orElseThrow(() -> new EntityNotFoundException("User not found"));
+
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
```
find shop cart by userId
find and remove cart item by `findByIdAndShoppingCartId`

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+        User user = userRepository.findByEmail(auth.getName())
+                .orElseThrow(() -> new RuntimeException("User not found"));
+
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
```
find shop cart by userId
find and update cart item by `findByIdAndShoppingCartId`

Elena-Bruyako replied:
fix comment
don't complicate your logic

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+    public ShoppingCartResponseDto getCartInfo() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = userRepository.findByEmail(authentication.getName())
+                .orElseThrow(() -> new RuntimeException("User not found"));
```
this is duplication of code

Elena-Bruyako reviewed (changes requested):
see comments

misterSokor committed:
Update src/main/java/com/store/litflix/service/impl/CartServiceImpl.java

I agree here, but I'm just a little concerned that returning the pre-saved shoppingCart might not reflect the final persisted state of the object , is that correct?

Co-authored-by: Olena Bruiako <bruyako.elena@gmail.com>

misterSokor committed:
Added validation annotations to request DTOs, renamed DTO classes for clarity, and improved cart item removal logic by verifying ownership via userId and shoppingCartId

Elena-Bruyako reviewed (changes requested):
fix all previous comments

misterSokor committed:
Ensure new carts have no preset ID, replace RuntimeException with EntityNotFoundException, rename and validate DTO fields, remove redundant variable, extract cart creation to dedicated method, and fetch users by ID.

Elena-Bruyako commented on src/main/java/com/store/litflix/controller/CartController.java:
```diff
@@ -0,0 +1,73 @@
+            description = "Adds a book to the shopping cart."
+                          + " If the book already exists, it increases the quantity."
+    )
+    @PostMapping
```
add status

Elena-Bruyako commented on src/main/java/com/store/litflix/dto/cart/CartItemRequestDto.java:
```diff
@@ -0,0 +1,9 @@
+
+@Data
+public class CartItemRequestDto {
+    private Long bookId;
```
should we validate request?

Elena-Bruyako commented on src/main/java/com/store/litflix/dto/cart/UpdateCartItemDto.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+public class UpdateCartItemDto {
+    @Min(value = 1, message = "Quantity must be at least 1")
```
```suggestion
    @Positive
```

Elena-Bruyako commented on src/main/java/com/store/litflix/mapper/BookMapper.java:
```diff
@@ -1,18 +1,29 @@
+    @Named("mapCategoryIds")
+    static List<Long> mapCategoryIds(Set<Category> categories) {
+        return categories == null ? null :
+                categories.stream().map(com.store.litflix.model.Category::getId).toList();
```
use import for Category
```suggestion
                categories.stream()
                .map(com.store.litflix.model.Category::getId)
                .toList();
```

Elena-Bruyako commented on src/main/java/com/store/litflix/model/ShoppingCart.java:
```diff
@@ -0,0 +1,36 @@
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
+    private Set<CartItem> cartItems = new HashSet<>();
```
add `isDeleted`

Elena-Bruyako commented on src/main/java/com/store/litflix/model/User.java:
```diff
@@ -43,6 +45,9 @@ public class User implements UserDetails {
     @Column(nullable = false)
     private boolean isDeleted;
 
+    @OneToOne(mappedBy = "user", cascade = CascadeType.ALL, orphanRemoval = true)
+    private ShoppingCart shoppingCart;
+
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/litflix/security/CustomUserDetailsService.java:
```diff
@@ -21,4 +21,9 @@ public UserDetails loadUserByUsername(String email)
                         new EntityNotFoundException("Can't find user by email"));
     }
+
+    public UserDetails loadByUserId(Long userId) {
```
why do you need this method?

misterSokor replied:
I use this method in JwtAuthenticationFilter , method doFilterInternal, line:   UserDetails userDetails =
                    ((CustomUserDetailsService)userDetailsService).loadByUserId(userIdFromToken);

Elena-Bruyako commented on src/main/java/com/store/litflix/security/JwtAuthenticationFilter.java:
```diff
@@ -30,8 +30,9 @@ protected void doFilterInternal(HttpServletRequest request,
         if (token != null && jwtUtil.isValidToken(token)) {
-            String username = jwtUtil.getUsernameFromToken(token);
-            UserDetails userDetails = userDetailsService.loadUserByUsername(username);
+            Long userIdFromToken = jwtUtil.getUserIdFromToken(token);
+            UserDetails userDetails =
+                    ((CustomUserDetailsService)userDetailsService).loadByUserId(userIdFromToken);
```
```suggestion
```

misterSokor replied:
It is a bit complicated to me. Are we switching to using email  in the token instead of userId? If so, I’ll switch to loadUserByUsername again. Is it correct?

Elena-Bruyako commented on src/main/java/com/store/litflix/security/JwtUtil.java:
```diff
@@ -24,9 +24,9 @@ public JwtUtil(@Value("${jwt.secret}") String secretString) {
     }
 
-    public String generateToken(String userNameAsEmail) {
+    public String generateToken(Long userId) {
```
```suggestion
    public String generateToken(String username) {
```

misterSokor replied:
I can't understand if I should use email or user id in the JVT token

misterSokor replied:
?

Elena-Bruyako commented on src/main/java/com/store/litflix/security/JwtUtil.java:
```diff
@@ -46,8 +46,8 @@ public boolean isValidToken(String token) {
-    public String getUsernameFromToken(String token) {
-        return getClaimFromToken(token, Claims::getSubject);
+    public Long getUserIdFromToken(String token) {
+        return Long.parseLong(getClaimFromToken(token, Claims::getSubject));
```
revert changes

misterSokor replied:
I can't understand if I should use email or user id in the JVT token

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+@Service
```
```suggestion
@Service
@Transactional
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+    public ShoppingCartResponseDto addBookToCart(CartItemRequestDto requestDto) {
+        User user = getCurrentUser();
+        Book book = bookRepository.findById(requestDto.getBookId()).orElseThrow(
+                () -> new EntityNotFoundException("The book was not found.")
```
add parameter to message

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+                () -> new EntityNotFoundException("The book was not found.")
+        );
+
+        int quantity = requestDto.getQuantity();
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(book);
+        cartItem.setQuantity(quantity);
```
```suggestion
        cartItem.setQuantity(requestDto.getQuantity());
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+        cartItemRepository.save(cartItem);
+
+        shoppingCart.getCartItems().add(cartItem);
+        user.setShoppingCart(shoppingCart);
```
remove 
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+
+    private User getCurrentUser() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = userRepository.findByEmail(authentication.getName())
```
provide Authentication for each endpoint in conroller and get `userId` for faster searching

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    private User getCurrentUser() {
```
private methods should be placed after all public

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+    private User getCurrentUser() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = userRepository.findByEmail(authentication.getName())
+                .orElseThrow(() -> new RuntimeException("User not found"));
```
don't throw RuntimeException
always add parameter to exception message

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+        int quantity = requestDto.getQuantity();
+
+        ShoppingCart shoppingCart = cartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createShoppingCart(user));
```
create shop cart during registration of user, remove this logic here

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+    public ShoppingCart createShoppingCart(User user) {
+        ShoppingCart cart = new ShoppingCart();
+        cart.setUser(user);
+        cart.setCartItems(new HashSet<>());
+
+        return cartRepository.save(cart);
```
```suggestion
        return cartRepository.save(cart);
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Override
+    public ShoppingCart createShoppingCart(User user) {
```
```suggestion
    public void createShoppingCart(User user) {
```

misterSokor replied:
I am not sure about making that method void, because that creates more quires to the DB

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+        User user = getCurrentUser();
+
+        ShoppingCart shoppingCart = cartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found"));
```
parameter

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+            throw new EntityNotFoundException("Cart item not found");
+        }
+
+        if (!cart.getUser().getId().equals(user.getId())) {
+            throw new RuntimeException(
+                    "You are not allowed to delete this cart item.");
+        }
+
```
you already find shop cart by particular user, redundant checking
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/litflix/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found"));
+
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId());
+        if (cartItem == null) {
```
could you use `orElseThrow()`?

Elena-Bruyako reviewed (changes requested):
see comments.
Please, be attentive

misterSokor committed:
Refactor security and cart services, simplify CartItem creation, clean up unused methods and interfaces, and add missing annotations

liliia-ponomarenko commented on src/main/java/com/store/litflix/controller/BookController.java:
```diff
@@ -66,8 +66,8 @@ public List<BookDto> searchBooks(@RequestParam(required = false) String[] title,
-    public BookDto updateBook(@PathVariable Long id,
-                              @RequestBody UpdateBookRequestDto requestDto) {
+    public BookResponseDto updateBook(@PathVariable Long id,
+                                      @RequestBody UpdateBookRequestDto requestDto) {
```
Valid annotation missed

liliia-ponomarenko commented on src/main/java/com/store/litflix/controller/CartController.java:
```diff
@@ -0,0 +1,88 @@
+        return cartService.addBookToCart(request, userId);
+    }
+
+    private Long extractUserId(Authentication authentication) {
+        return Long.parseLong(authentication.getName());
+    }
```
Move private method after public

liliia-ponomarenko commented on src/main/java/com/store/litflix/model/CartItem.java:
```diff
@@ -0,0 +1,36 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotNull
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @NotNull
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
    private ShoppingCart shoppingCart;
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
    private Book book;
```

liliia-ponomarenko reviewed (changes requested):
Good job! A few comments ;)

misterSokor committed:
Refactored controller and model: updated return type and added @Valid in updateBook, reordered extractUserId method, and cleaned up CartItem annotations.

liliia-ponomarenko reviewed (approved):
Good job!
